cdap-master/src/main/java/co/cask/cdap/data/runtime/main/StartupChecker.java

```diff
@@ -0,0 +1,179 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data.runtime.main;
+
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.startup.ConfigurationLogger;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.google.inject.Inject;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.hbase.client.HConnection;
+import org.apache.hadoop.hbase.client.HConnectionManager;
+import org.apache.hadoop.yarn.client.api.YarnClient;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Performs some sanity checks to catch any glaring problems that would prevent master from starting up.
+ */
+public class StartupChecker extends AbstractIdleService {
+  private static final Logger LOG = LoggerFactory.getLogger(StartupChecker.class);
+  private final Configuration hConf;
+  private final CConfiguration cConf;
+  private final LocationFactory locationFactory;
+  // used because some operations hang if the dependent service is not running.
+  private final ExecutorService executorService;
+
+  @Inject
+  StartupChecker(Configuration hConf, CConfiguration cConf, LocationFactory locationFactory) {
+    this.hConf = hConf;
+    this.cConf = cConf;
+    this.locationFactory = locationFactory;
+    this.executorService = Executors.newSingleThreadExecutor(
+      new ThreadFactoryBuilder().setNameFormat("startup-checker").build());
+  }
+
+  /**
+   * Run sanity checks to see if there are any glaring problems that will prevent master from starting up.
+   * These are the types of things that are not likely to be resolved without manual intervention, such as incorrect
+   * directory permissions, wrong hive dependencies, and underlying infrastructure like HDFS being unavailable.
+   *
+   * Other types of errors that may eventually resolve themselves, such as failure to connect to transaction service,
+   * are handled in a separate method.
+   *
+   * This method is a no-op if {@link Constants#STARTUP_CHECKS_ENABLED} is false.
+   *
+   * @return whether the sanity checks succeeded or not
+   */
+  public boolean runSanityChecks() {
+    boolean noErrors = true;
+
+    ConfigurationLogger.logImportantConfig(cConf);
```

Comment:
Should we print number of open file handles currently configured. Also, should we check if zookeeper is up and running ?